peps/pep-0734.rst

```diff
@@ -0,0 +1,800 @@
+types like mutexes.  The ``threading`` module exposes several.
+However, interpreters cannot share objects which means they cannot
+share ``threading.Lock`` objects.
+
+The ``interpreters`` module does not provide any such dedicated
+synchronization primitives.  Instead, ``Queue`` objects provide
+everything we need.
+
+For example, if there's a shared resource that needs managed
+access then a queue may be used to manage it, where the interpreters
+pass an object around to indicate who can use the resource::
+
+   import interpreters
+   from mymodule import load_big_data, check_data
+
+   numworkers = 10
+   control = interpreters.create_queue()
+   data = memoryview(load_big_data())
+
+   def worker():
+       interp = interpreters.create()
+       interp.prepare___main__(control=control, data=data)
+       interp.exec("""if True:
+           from mymodule import edit_data
+           while True:
+               token = control.get()
+               edit_data(data)
+               control.put_nowait(token)
+           """)
+   threads = [threading.Thread(target=worker) for _ in range(numworkers)]
+   for t in threads:
+       t.start()
+
+   token = 'football'
+   control.put_nowait(token)
+   while True:
+       control.get()
+       if not check_data(data):
+           break
+       control.put_nowait(token)
+
+If an interpreter needs to wait for another to reach a certain point
+then they can synchronize through a queue.  We just need to make sure
+to synchronize in both directions::
+
+   import interpreters
+
+   sync = interpreters.create_queue()
+
+   def worker():
+       interp = interpreters.create()
+       interp.prepare___main__(sync=sync)
+       interp.exec("""if True:
+           # do step 1
+           token = sync.get()
+           sync.put(token)
+           # do step 2
+           sync.put(None)
+           sync.get()
+           # do step 3
+           sync.put(None)
+           sync.get()
+           """)
+   t = threading.Thread(target=worker)
+   t.start()
+
+   # do step A
+   sync.put(None)
+   sync.get()
+   # do step B
+   token = sync.get()
+   sync.put(token)
+   # do step C
+   token = sync.get()
+   sync.put(token)
+
+Exceptions
+----------
+
+* ``RunFailedError``
+      Raised from ``Interpreter.exec()`` when there's an uncaught
+      exception.
+
+      This exception is a subclass of ``RuntimeError``.
+
+* ``QueueEmpty``
+      Raised from ``Queue.get()`` (or ``get_nowait()`` with no default)
+      when the queue is empty.
+
+      This exception is a subclass of ``queue.Empty``.
+
+* ``QueueFull``
+      Raised from ``Queue.put()`` (with a timeout) or ``put_nowait()``
+      when the queue is already at its max size.
+
+      This exception is a subclass of ``queue.Full``.
+
+InterpreterPoolExecutor
+-----------------------
+
+The new ``concurrent.futures.InterpreterPoolExecutor`` will be
+a subclass of ``concurrent.futures.ThreadPoolExecutor``, where each
```

Comment:
Maybe just promise `c.f.Executor`? That's an abstract type that defines the interface.